mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -108,34 +121,25 @@ impl PeerManager {
     }
 
     // Returns true if `peer` is discovered for the first time, else false.
-    pub fn add_peer(&self, peer: PeerNetworkId, origin: ConnectionOrigin) -> bool {
+    pub fn add_peer(&self, peer: PeerNetworkId, metadata: ConnectionMetadata) -> bool {
         let mut peer_states = self.peer_states.lock();
         let is_new_peer = !peer_states.contains_key(&peer);
-        if self.is_upstream_peer(&peer, Some(origin)) {
+        if self.is_upstream_peer(&peer, Some(metadata.origin)) {
             counters::active_upstream_peers(&peer.raw_network_id()).inc();
             if peer.raw_network_id() == NetworkId::Validator {
                 // For a validator network, resume broadcasting from previous state.
                 // We can afford to not re-broadcast here since the transaction is already in a validator.
-                peer_states
-                    .entry(peer)
-                    .or_insert(PeerSyncState {
-                        timeline_id: 0,
-                        is_alive: true,
-                        broadcast_info: BroadcastInfo::new(),
-                    })
-                    .is_alive = true;
+                if is_new_peer {
+                    peer_states.insert(peer, PeerSyncState::new(metadata));
+                } else if let Some(peer_state) = peer_states.get_mut(&peer) {
+                    peer_state.is_alive = true;
+                    peer_state.metadata = metadata;
+                }
             } else {
                 // For a non-validator network, potentially re-broadcast any transactions that have not been
                 // committed yet.
                 // This is to ensure better reliability of transactions reaching the validator network.
-                peer_states.insert(
-                    peer,
-                    PeerSyncState {
-                        timeline_id: 0,
-                        is_alive: true,
-                        broadcast_info: BroadcastInfo::new(),
-                    },
-                );
+                peer_states.insert(peer, PeerSyncState::new(metadata));
```

Comment:
why isn't it possible for this to be an update?
the notion that we can get here feels like all peers should be "new" and that we should remove their state when they disconnect... since we cannot guarantee their state if they have crashed and we don't have some form of pre-flight to see what they have (e.g., exchanging a bloomfilter or something)

Reply:
Yeah, I don't have good reasoning here other than the comments that are already here.  If you see below in the `disable_peer` I added a comment about garbage collection.

In theory, if a large number of peers were to connect over a long period of time, this map would grow greatly, as there *never* is a time we remove from the map.